text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+
+Each TokenType has an explicit minimum fee specified by node operators, and included in the attestation handshake, ensuring that all nodes in the network are configured with the same fee minimums and TokenType sets. New TokenTypes can be added only with unanimous agreement from all node operators. 
+
+A transaction specifies the TokenType for the entire transaction, and includes a "Proof of Opening" that all `token_ids` in the inputs and outputs are the same. This prevents spending one TokenType as a member of another TokenType. The TransactionBuilder now takes a `token_id` as an argument in its initializer. The TokenType is set in the clear on the TxPrefix for a transaction, which is only accessed in the enclave, after the payload is delivered via an encrypted, attested channel. This is similar to the TxInputs, and thus has a similar threat model for confidentiality. See [Confidentiality Analysis](#confidentiality-analysis) for an assessment of risk.
+
+Clients must choose how to expose multiple asset types to their users. There may be clients who wish to only expose a subset of asset types, or clients who wish to provide balances and transactionality in all asset types. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Masked Token ID Field in the TXO
+
+A `TokenType` is added which specifies a `u32 token_id`. The `token_id` for each TXO represents the Token Type of that TXO, with the first `TokenType` being `MOB = token_id(0)`. New TokenTypes increase the `token_id` monotonically. For example, the `TokenType` struct may be something like,
+
+```
+struct TokenType {
+
+    \\\ Token Name, such as MOB
+    string token_name,
+    
+    \\\ Token ID to be represented confidentially in each TXO
+    u32 token_id,
+    
+    \\\ Curve base point, unique for each TokenType, from which the TXO's commitment is 
+    \\\ generated with the value of the `amount` in the TXO
+    RistrettoPoint H_i
+}
+```
+
+The `masked_token_id` field is added to the [`Amount`](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/core/src/amount/mod.rs#L28) in each TXO, and functions similar to `masked_value`. The protocol representation of the TXO uses Protobuf, which enables backward compatability in this way: if the field is not populated (Protobuf bytes field of length 0), the default `token_id` is 0. Otherwise, the field is a Protobuf bytes field of length 4. Please see the [Upgrade Plan](#upgrade-plan) for more discussion on rolling out this change with consideration for backward compatibility.
+
+- To compute the 4 byte `masked_token_id` from the `token_id`, we hash the TXO shared secret, with a prefix, XORed with the `token_id`, and take the little endian representation of those bytes.
+
+    ```
+    masked_token_id = (token_id ^ Blake2B(token_id_tag | shared_secret)).to_le()
```

Comment:
`token_id_tag` is meant to be a domain tag - some string, and | was meant to denote concatenation - definitely open to sugs for fixing up notation for clarity!

Reply:
`token_id_tag` is meant to be a domain tag - some string, and | was meant to denote concatenation - definitely open to sugs for fixing up notation for clarity!

Reply:
Maybe instead of `token_id_tag` just call it `domain_tag` and the reader will understand?

I think `||` is more commonly used to denote concatenation in these kinds of docus, that would be my sug